rfc/rfc-78/rfc-78.md

```diff
@@ -0,0 +1,339 @@
+- Rewrite LSM based timeline to archived timeline. we have to deduce writer properties and introduce boundaries based on that (i.e. until which commit to archive).
+- We have two options wrt handling active timeline instants(for eg, pending clustering instants):
+   A: No changes to active instants. In order to support 1.x tables w/ 0.16.0 reader, already these changes might have been ported to 0.16.0 reader. But we might have to ensure downgrade from 0.16.0 to any older version of hudi does take care of rewriting the active timeline instants.
+   B: While downgrading from 1.x to 0.16.0, let’s rewrite the active timeline instants too. So that after downgrade, we can't differentiate whether table was natively created in 0.16.0 or was it upgraded to 1.x and then later downgraded. The con in Option A is not an issue here since we take care of active timeline instants during 1.x to 0.16.0 downgrade only. So, downgrade from 0.16.0 to any lower hudi version does not need to do any special handling if table was native to 0.16.0 or was it downgraded from 1.x to 0.16.0
+   Our proposal is to go with Option B to ensure 0.16.0 table will be intact and not have any residues from 1.x writes. While the rewrite of active instants are happening, we could see some read failures. But 
+   we do not want to have a 0.16.0 (post upgrade) table to deal w/ 1.x timeline intricacies. 
+- If there are new MDT partitions (functional index, sec index), nuke them. Update table properties.
+- Check latest snapshot files for log files w/ new features enabled. For eg, deletion vector, partial update. If any log files are found in latest snapshot, we have to trigger compaction for those file groups. This calls for a custom compaction strategy as well which compacts only for certain file groups.
+   - In order to reduce downtime to downgrade, we are inclining to do this only for snapshot reads. If there are older file slices and users are issuing time travel or incremental reads, and if they have any of the new features enabled (deletion vector, partial update), readers could break after downgrade.
+
+## Implementation details
+
+To ensure read capability of 1.x tables in 0.16.0, documenting the changes required at class level. Other two sections (0.16.0 upgrade to 1.x or 1.x downgrade to 0.16.0) should be fairly straightforward and so not calling out the obvious.
+
+### Timeline changes:
+Let’s reiterate what we need to support w/ 0.16.0 reader.
+
+### Timeline read of 1.x need to be supported
+- Commit instants w/ completion time should be readable. HoodieInstant parsing logic to parse completion time should be ported over.
+- We could ignore the completion time log file ordering semantics since we don’t plan to support it yet in 0.16.0. But reading should not break.
+- Pending Clustering commits using “cluster” action should be readable in 0.16.0 reader.
```

Comment:
NTR: 
As as optimization, we are planning to consider "table upgrade commit time" to help us here. because, just after upgrading to 1.x, we have to consider both pending replace commits and pending clustering commits (for clustering operations). But at some point eventually, we want to sunset checking replace commit timeline and only consider clustering timeline (1 month or 2 months or next release 1.0.1). But until then, this config "table upgrade commit time" will help us do deterministic processing. for eg, once active timeline's first entry is > "table upgrade commit time", its safe to consider only pending clustering timeline and ignore pending replace timeline.